### ISSUE TITLE
Discuss Lerna styling approaches for docs

### ISSUE THREAD

**Issue by exogen** [Haaretz/htz-frontend#36]
For the upcoming documentation, I'm going to include our recommended approach to doing various things with Fela. There are several ways for components to end up with the right `className` using `react-fela`. I'd like to get thoughts on @TxHawks' preferences for the different approaches.

Let's use the current `Logo` implementation as an example. It's currently written in a way that I think betrays my inexperience with Fela, but let's go over it first anyway:

### Direct `renderer` access

This requires setting `contextTypes`, which is a bummer.

```jsx
const style = () => { /* a Fela rule, omitted in remaining examples. */ };

const Logo = (props, { renderer, }) => {
  const className = renderer.renderRule(style, props);
  return (
    <svg className={className} ...snip...>
      <path d="...snip..." />
      <path d="...snip..." />
    </svg>
  );
};

Logo.contextTypes = { renderer: PropTypes.object, };
```

### Easier `renderer` access via a HoC

In order to provide a nicer API when direct `renderer` access is necessary, it's typical to make a HoC so that components can access things from `props` instead. That way people don't have to worry or even know about `contextTypes`. That would look like this (roughly):

```jsx
// This would live in `htz-components`:
function withRenderer(Component) {
  function WithRenderer(props, { renderer }) {
    return <Component renderer={renderer} {...props} />
  }
  WithRenderer.contextTypes = { renderer: PropTypes.object, };
  return WithRenderer;
}

// Use like so:
const Logo = withRenderer((props) => {
  const className = props.renderer.renderRule(style, props);
  return (
    <svg className={className} ...snip...>
      <path d="...snip..." />
      <path d="...snip..." />
    </svg>
  );
});
```

### Using `createComponent`

I didn't want to use the `createComponent` API for the `Logo` class because I didn't like having the intermediate single-use-only `LogoSvg` component. That would look like this:

```jsx
const LogoSvg = createComponent(style, 'svg', ['viewBox', 'width', 'height']);

const Logo = (props) => {
  // Note, don't have to worry about `renderer`.
  return (
    <LogoSvg {...props} ...snip...>
      <path d="...snip..." />
      <path d="...snip..." />
    </LogoSvg>
  )
}
```

### Using `connect`

I prefer this approach to both `createComponent` and `withRenderer`.

```jsx
const Logo = ({ styles }) => {
  return (
    <svg className={styles.container} ...snip...>
      <path d="...snip..." />
      <path d="...snip..." />
    </svg>
  );
}

export default connect({ container: style })(Logo);
```

**Comment by exogen** [comment]
It will also be worth documenting what your preferred naming and `export` conventions are in this scenario. When you wrap something with an HoC, there are actually two components. In the `connect` and `withRenderer` examples:

```jsx
const Logo = (props) => { ... };

const StyledLogo = connect({ ... })(Logo);
// ...or...
const LogoWithRenderer = withRenderer(Logo);

// What to export?

export Logo;
export default StyledLogo;
```

Splitting these up into multiple files is overkill, IMO. And there will be cases where you'll have multiple HoCs wrapping a component.

Conventions you'll want to establish:

1. Do you export the underlying, undecorated component? e.g. `export Logo`, but leave the `default` export the decorated one? (The underlying component would likely never actually be used, but exporting it can be useful for testing purposes.)
2. Should you try to give the decorated components unique names, reassign them to `Logo`, or skipping naming altogether and `export default connect({ ... })(Logo);`.

**Comment by TxHawks** [comment]
#### Creating components 
I'd rather avoid the directly using `renderer` and the `HoC` approaches. They seem messy and less scaleable, Especially directly accessing `renderer`.

[kilvin](https://github.com/rofrischmann/kilvin), which are maintained by the creator of Fela, and [cf-ui](https://github.com/cloudflare/cf-ui/tree/master/packages), which is the biggest component library using Fela opted for the `createComponent` approach (`cf-ui` modified it a bit by wrapping it with their own component, but it's mostly cosmetic.

As for `connect`, there is this small [helper](https://github.com/dustin-H/fela-styles-connector) for it, which is nice.

I think the right approach is to use `createComponent` for anything that has children, and `connect` for anything that's one-off.

#### Naming
I think we should plain forbid directly using `renderer` and discourage the `HoC` approach. 

I think the more common pattern for using `createComponent` is something like this:
```js
const styles = () => { /* rules */ };

function Logo ({ className, ...snip }) {
  return (
    <svg className={className} ...snip..>
      <path d="...snip..." />
      <path d="...snip..." />
    </svg>
  );
}
const styledLogo = createComponent(styles, Logo);
```

`componentName`  -->  `styledComponentName` is a good convention for both the `createElement` and `connect`.

#### Export
I say export all for testing, but only expose the styled-component to `index.js` leaving it outside the documented API

Makes sense?

**Comment by TxHawks** [comment]
On a semi-related note, I think every app will probably need its own `renderer` and perhaps `StyleProvider` in order to support easily reusing components between RTL <-> LTR contexts. 

I'll open a PR for that in a bit

**Comment by exogen** [comment]
I don't have much experience with handling RTL concerns, hoping you can elaborate on why this would affect the `renderer` and `StyleProvider`.

**Comment by TxHawks** [comment]
I'll push a commit in a second and explain

**Comment by TxHawks** [comment]
The [`HaaretzDocument` class](https://github.com/Haaretz/htz-frontend/blob/master/packages/components/htz-components/src/document.js#L13-L14) in `htz-components` uses the renderer from an import, meaning it can not be defined on the app level. 
Can this be changed so that the renderer can be defined on the app level?

**Comment by exogen** [comment]
Yes, we can do that. In order to not have to duplicate the document code for every app, we'll have to make the thing in `htz-components` something like a `createDocument` function that accepts the renderer as an argument.

**Comment by TxHawks** [comment]
Wanted to avoid that, but that works

**Comment by kedarguy** [comment]
Hey,

So me and @TawKhoury  Played around with Fela for the last week and i'll try and document the main issues we ran into so it will be easier for the other guys to start working with Fela + try to contribute to figuring out the best practice that we will want to use.

I got some examples I think will be helpful on 'htz-component-testing-fela' branch, in the playground components we built, PlaygroundA, PlayfroundB and PlaygroundC components

main questions we have:

Is there something in fela like combineRules that will take in a 
component and combine its rules with new one and overwrite overlapping rules?

The purpose is to get a component that already has style and add or overwrite some of the css (adding is not a problem, but we didnt find a solution for overwriting )

[htz-component-testing-fela](https://github.com/Haaretz/htz-frontend/tree/htz-component-testing-fela)

We explained in the code comments the different issues we ran into.
